Adds s2n_connection section to usage guide (#3605)
aws/s2n-tls: docs/USAGE-GUIDE.md

The TLS handshake requires a significant amount of memory, but once a connection is established the handshake memory can be freed by calling `s2n_connection_free_handshake()`. This function should be called after the handshake is successfully negotiated and logging or recording of handshake data is complete.

The input and output buffers consume the most memory on the connection after the handshake. It may not be necessary to keep these buffers allocated when a connection is in a keep-alive or idle state. Call `s2n_connection_release_buffers()` to wipe and free the `in` and `out` buffers associated with a connection to reduce memory overhead of long-lived connections.

### Connection Reuse
Connection objects should be re-used across many connections (to avoid excessive de-allocating and allocating memory). Calling `s2n_connection_wipe()` will wipe an individual connection's state and allow the connection object to be re-used for a new TLS connection.

### Connection Info

> Use the connection getter methods to interrogate the connection object about the TLS handshake. `s2n_connection_get_client_protocol_version()` and `s2n_connection_get_server_protocol_version()` return the protocol version number supported by the client or the server respectively. `s2n_connection_get_actual_protocol_version()` returns the protocol version actually negotiated by the handshake. `s2n_connection_get_cipher()` returns the cipher suite selected by the handshake. More connection getters can be found in the [doxygen guide](https://aws.github.io/s2n-tls/doxygen/).

Reviewer: I'd take a look at https://github.com/aws/s2n-tls/issues/1744 Josh makes a very good point, this is super confusing.

Maybe instead of including them in the paragraph, we try something like:

```suggestion
s2n-tls provides many methods to retrieve details about the handshake and connection, such as the parameters negotiated with the peer. For a full list, see [doxygen guide](https://aws.github.io/s2n-tls/doxygen/).

#### Protocol Version
s2n-tls provides multiple different methods to get the protocol version. They will not always agree, and it's important to call the correct one for your use case.
* `actual_protocol_version`: The actual protocol version negotiated during the handshake. This is the primary value referred to as "protocol_version", and the most commonly used.
* `server_protocol_version`: ...etc...
```

If you are trying to use FIPS mode, you must enable FIPS in your libcrypto library (probably by calling `FIPS_mode_set(1)`) before calling `s2n_init()`.

## Connection

> 
> Users will need to create a `s2n_connection` struct to store all of the state necessary to negotiate a TLS connection. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed.

Reviewer: Oh, sorry I didn't catch this the first time. "negotiate" might be misleading, since we need to keep it around after the negotiation / handshake.
```suggestion

Users will need to create a `s2n_connection` struct to store all of the state necessary for a TLS connection. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed.
```

### Connection Memory

The connection struct is roughly 4KB with some variation depending on how it is configured. Maintainers of the s2n-tls library carefully consider increases to the size of the connection struct as they are aware some users are memory-constrained.

The TLS handshake requires a significant amount of memory, but if necessary the handshake memory can be freed by calling `s2n_connection_free_handshake()`. This function should be called after the handshake is successfully negotiated and logging or recording of handshake data is complete. This function should not be called if you're planning to reuse the connection object for multiple connections. See [connection reuse](#connection-reuse).

> 

Reviewer: I'm still not sure this is quite right. Using s2n_connection_wipe is the most common reason you wouldn't call s2n_connection_free_handshake, but you can still call it if you're planning to reuse the connection-- maybe connections are pooled but also very long-lived?

Maybe:
> An s2n-tls connection allocates memory for the TLS handshake that isn't required when later sending and receiving application data. This memory can be freed by calling `s2n_connection_free_handshake`, but some functions that print information about the handshake may not produce meaningful results after the handshake memory is freed. Applications may want to free handshake memory due to memory constraints, particularly if connections are long-lived. However, the handshake memory can be reused for another connection if `s2n_connection_wipe` is called, so freeing it may later result in more potentially expensive memory allocations.

Author: Didn't copy this paragraph word for word, but I did pull a lot from it.

> A connection struct has memory allocated specifically for the TLS handshake. Memory-constrained users can free that memory by calling `s2n_connection_free_handshake()`, after the handshake is successfully negotiated. Note that the handshake memory can be reused for another connection if `s2n_connection_wipe()` is called, so freeing it may result in more memory allocations later. Additionally some functions that print information about the handshake may not produce meaningful results after the handshake memory is freed.

Reviewer: ```suggestion
A connection struct has memory allocated specifically for the TLS handshake. Memory-constrained users can free that memory by calling `s2n_connection_free_handshake()` after the handshake is successfully negotiated. Note that the handshake memory can be reused for another connection if `s2n_connection_wipe()` is called, so freeing it may result in more memory allocations later. Additionally some functions that print information about the handshake may not produce meaningful results after the handshake memory is freed.
```
